File: riveredge-frontend/src/apps/kuaizhizao/services/purchase.ts
/**
 * 采购订单API服务
 *
 * 提供采购订单相关的API调用接口
 *
 * @author RiverEdge Team
 * @date 2025-01-01
 */

import { apiRequest } from '../../../services/api';

const PURCHASE_ORDER_LIST_LIMIT_MAX = 100;

function clampPurchaseOrderListLimit(limit: unknown): number | undefined {
  if (typeof limit !== 'number' || Number.isNaN(limit)) return undefined;
  return Math.max(1, Math.min(PURCHASE_ORDER_LIST_LIMIT_MAX, Math.trunc(limit)));
}

export interface ActionCapability {
  allowed: boolean;
  reason?: string | null;
}

export interface PurchaseOrderCapabilities {
  update?: ActionCapability;
  delete?: ActionCapability;
  submit?: ActionCapability;
  withdraw_submit?: ActionCapability;
  approve?: ActionCapability;
  revoke_approval?: ActionCapability;
  push_receipt_notice?: ActionCapability;
  print?: ActionCapability;
}

/**
 * 采购订单接口定义
 */
export interface PurchaseOrder {
  id?: number;
  tenant_id?: number;
  order_code?: string;
  supplier_id?: number;
  supplier_name?: string;
  supplier_contact?: string;
  supplier_phone?: string;
  order_date?: string;
  delivery_date?: string;
  order_type?: string;
  total_quantity?: number;
  total_amount?: number;
  tax_rate?: number;
  tax_amount?: number;
  net_amount?: number;
  currency?: string;
  exchange_rate?: number;
  status?: string;
  source_type?: string;
  source_id?: number;
  review_status?: string;
  reviewer_id?: number;
  reviewer_name?: string;
  review_time?: string;
  review_remarks?: string;
  buyer_id?: number;
  buyer_name?: string;
  notes?: string;
  items_count?: number;
  created_at?: string;
  updated_at?: string;
  items?: PurchaseOrderItem[];
  /** 变更原因 (V2 审计) */
  change_reason?: string;
  fee_details?: any[];
  total_fee_amount?: number;
  capabilities?: PurchaseOrderCapabilities;
}

export interface PurchaseOrderItem {
  id?: number;
  material_id?: number;
  material_code?: string;
  material_name?: string;
  material_spec?: string;
  ordered_quantity?: number;
  unit?: string;
  unit_price?: number;
  total_price?: number;
  received_quantity?: number;
  outstanding_quantity?: number;
  required_date?: string;
  actual_delivery_date?: string;
  source_id?: number;
  notes?: string;
  /** 分摊落地成本 (V2) */
  landing_cost?: number;
  /** 杂费明细 (V2) */
  additional_fees_details?: LandingCostFeeItem[];
  capabilities?: PurchaseOrderCapabilities;
}

export interface PurchaseOrderListParams {
  skip?: number;
  limit?: number;
  supplier_id?: number;
  status?: string;
  review_status?: string;
  order_date_from?: string;
  order_date_to?: string;
  delivery_date_from?: string;
  delivery_date_to?: string;
  buyer_id?: number;
  keyword?: string;
}

export interface PurchaseOrderListResponse {
  data: PurchaseOrder[];
  total: number;
  success: boolean;
}

/** 采购入库选单弹窗候选项 */
export interface PurchaseReceiptPullCandidate {
  id: number;
  order_code: string;
  supplier_name?: string;
  status: string;
  order_date?: string;
  delivery_date?: string;
  items_count?: number;
  ordered_total?: number;
  received_total?: number;
  outstanding_total?: number;
  pullable: boolean;
  lifecycle?: {
    current_stage_name?: string;
    sub_stages?: Array<{ key: string; label: string; status: string }>;
  };
}

export interface PurchaseReceiptPullCandidateListResponse {
  data: PurchaseReceiptPullCandidate[];
  total: number;
  success: boolean;
}

/** 采购订单统计（用于指标卡片） */
export interface PurchaseOrderStatistics {
  active_count: number;
  pending_review_count: number;
  in_progress_count: number;
  overdue_count: number;
  total_amount: number;
  // 新增指标
  monthly_arrival_rate?: number;
  annual_total_amount?: number;
  supplier_on_time_rate?: number;
  // 趋势数据
  trends?: {
    overdue?: number[];
    arrival_rate?: number[];
    annual_total?: number[];
    efficiency?: number[];
  };
  /** 效率同比（百分比） */
  efficiency_yoy?: number;
  /** 年度总额同比（百分比，如 12.5 表示 12.5%） */
  annual_total_yoy?: number;
}

/** 获取采购订单统计 */
export async function getPurchaseOrderStatistics(): Promise<PurchaseOrderStatistics> {
  return apiRequest<PurchaseOrderStatistics>('/apps/kuaizhizao/purchase-orders/statistics', {
    method: 'GET',
  });
}

/**
 * 获取采购订单列表
 */
export async function listPurchaseOrders(params: PurchaseOrderListParams = {}): Promise<PurchaseOrderListResponse> {
  const limit = clampPurchaseOrderListLimit(params.limit);
  const safeParams = limit != null ? { ...params, limit } : params;
  return apiRequest<PurchaseOrderListResponse>('/apps/kuaizhizao/purchase-orders', {
    method: 'GET',
    params: safeParams,
  });
}

/** 采购入库选单弹窗：单次请求返回订单及入库数量汇总 */
export async function listPurchaseReceiptPullCandidates(params: {
  skip?: number;
  limit?: number;
  keyword?: string;
} = {}): Promise<PurchaseReceiptPullCandidateListResponse> {
  const limit = typeof params.limit === 'number'
    ? Math.max(1, Math.min(200, Math.trunc(params.limit)))
    : 100;
  return apiRequest<PurchaseReceiptPullCandidateListResponse>(
    '/apps/kuaizhizao/purchase-orders/receipt-pull-candidates',
    {
      method: 'GET',
      params: { ...params, limit },
    },
  );
}

/**
 * 获取采购订单详情
 */
export async function getPurchaseOrder(id: number): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/apps/kuaizhizao/purchase-orders/${id}`, {
    method: 'GET',
  });
}

/**
 * 创建采购订单
 */
export async function createPurchaseOrder(data: Partial<PurchaseOrder>): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>('/apps/kuaizhizao/purchase-orders', {
    method: 'POST',
    data,
  });
}

/**
 * 更新采购订单
 */
export async function updatePurchaseOrder(id: number, data: Partial<PurchaseOrder>): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/apps/kuaizhizao/purchase-orders/${id}`, {
    method: 'PUT',
    data,
  });
}

/**
 * 删除采购订单
 */
export async function deletePurchaseOrder(id: number): Promise<void> {
  return apiRequest<void>(`/apps/kuaizhizao/purchase-orders/${id}`, {
    method: 'DELETE',
  });
}

/**
 * 审核采购订单
 */
/**
 * 审核采购订单请求接口
 */
export interface PurchaseOrderApproveRequest {
  approved: boolean;
  review_remarks?: string;
}

/**
 * 审核采购订单
 */
export async function approvePurchaseOrder(id: number, data: PurchaseOrderApproveRequest): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/apps/kuaizhizao/purchase-orders/${id}/approve`, {
    method: 'POST',
    data,
  });
}

/**
 * 确认采购订单
 */
export async function confirmPurchaseOrder(id: number): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/apps/kuaizhizao/purchase-orders/${id}/confirm`, {
    method: 'POST',
  });
}

/**
 * 提交采购订单（非审核）
 */
export async function submitPurchaseOrder(id: number): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/apps/kuaizhizao/purchase-orders/${id}/submit`, {
    method: 'POST',
  });
}

/** 撤回提交（待审核 -> 草稿），走统一审核入口 */
export async function withdrawPurchaseOrder(id: number): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/core/uni-audit/purchase_order/${id}/withdraw`, {
    method: 'POST',
  });
}

/** 撤销审核（已确认/已驳回 -> 待审核），走统一审核入口 */
export async function revokePurchaseOrder(id: number): Promise<PurchaseOrder> {
  return apiRequest<PurchaseOrder>(`/core/uni-audit/purchase_order/${id}/revoke`, {
    method: 'POST',
  });
}

/**
 * 下推采购入库预览（返回批号等，供弹窗展示）
 */
export async function pushPurchaseOrderToReceiptPreview(
  id: number,
  receiptQuantities?: Record<number, number>
): Promise<{ items: Array<{ item_id: number; material_code: string; material_name: string; receipt_quantity: number; batch_number?: string }> }> {
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${id}/push-to-receipt-preview`, {
    method: 'POST',
    data: receiptQuantities || {},
  });
}

export type PushPurchaseOrderToReceiptOptions = {
  warehouseId?: number;
  lineWarehouses?: Record<number, number>;
  lineLocationIds?: Record<number, number>;
  lineLocationCodes?: Record<number, string>;
};

/**
 * 下推到采购入库
 */
export async function pushPurchaseOrderToReceipt(
  id: number,
  receiptQuantities?: Record<number, number>,
  batchNumbers?: Record<number, string>,
  options?: number | PushPurchaseOrderToReceiptOptions,
): Promise<any> {
  const opts: PushPurchaseOrderToReceiptOptions =
    typeof options === 'number' ? { warehouseId: options } : (options ?? {});
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${id}/push-to-receipt`, {
    method: 'POST',
    data: {
      receipt_quantities: receiptQuantities || {},
      ...(batchNumbers && Object.keys(batchNumbers).length > 0 ? { batch_numbers: batchNumbers } : {}),
      ...(opts.warehouseId != null && opts.warehouseId > 0 ? { warehouse_id: opts.warehouseId } : {}),
      ...(opts.lineWarehouses && Object.keys(opts.lineWarehouses).length > 0
        ? { line_warehouses: opts.lineWarehouses }
        : {}),
      ...(opts.lineLocationIds && Object.keys(opts.lineLocationIds).length > 0
        ? { line_location_ids: opts.lineLocationIds }
        : {}),
      ...(opts.lineLocationCodes && Object.keys(opts.lineLocationCodes).length > 0
        ? { line_location_codes: opts.lineLocationCodes }
        : {}),
    },
  });
}

/**
 * 下推到收货通知
 */
export async function pushPurchaseOrderToReceiptNotice(id: number, noticeQuantities?: Record<number, number>): Promise<any> {
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${id}/push-to-receipt-notice`, {
    method: 'POST',
    data: noticeQuantities || {},
  });
}

/**
 * 下推到采购发票
 */
export async function pushPurchaseOrderToInvoice(id: number): Promise<any> {
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${id}/push-to-invoice`, {
    method: 'POST',
  });
}

export async function pushPurchaseOrderToPurchaseReturn(data: {
  purchase_order_id: number;
  warehouse_id: number;
  warehouse_name?: string;
  return_quantities?: Record<number, number>;
}): Promise<{ id: number; return_code?: string }> {
  return apiRequest<{ id: number; return_code?: string }>('/apps/kuaizhizao/purchase-returns/pull-from-purchase-order', {
    method: 'POST',
    data,
  });
}

export async function pullPurchaseOrderFromInquiry(data: {
  inquiry_id: number;
  item_ids?: number[];
  persist_default_supplier_to_material?: boolean;
}): Promise<{ purchase_orders?: Array<{ purchase_order_id: number; purchase_order_code: string; supplier_id: number }> }> {
  return apiRequest('/apps/kuaizhizao/purchase-orders/pull-from-inquiry', {
    method: 'POST',
    data,
  });
}

/** 物料价格历史响应接口 */
export interface MaterialPriceHistoryResponse {
  material_id: number;
  history_items: Array<{
    order_id: number;
    order_code: string;
    order_date: string;
    supplier_id: number;
    supplier_name: string;
    unit_price: number;
    currency: string;
  }>;
  average_price: number;
  min_price: number;
  max_price: number;
}

/** 采购追踪响应接口 */
export interface PurchaseTrackingResponse {
  order_id: number;
  order_code: string;
  overall_progress: number;
  nodes: Array<{
    node_name: string;
    status: string;
    time?: string;
    operator?: string;
    detail?: string;
    is_completed: boolean;
    is_warning: boolean;
  }>;
}

/** 获取物料历史成交价 */
export async function getMaterialPriceHistory(materialId: number): Promise<MaterialPriceHistoryResponse> {
  return apiRequest<MaterialPriceHistoryResponse>(`/apps/kuaizhizao/material-price-history/${materialId}`, {
    method: 'GET',
  });
}

/** 获取采购订单履约追踪 */
export async function getPurchaseOrderTracking(orderId: number): Promise<PurchaseTrackingResponse> {
  return apiRequest<PurchaseTrackingResponse>(`/apps/kuaizhizao/purchase-orders/${orderId}/tracking`, {
    method: 'GET',
  });
}

/** 一键催单 */
export async function expeditePurchaseOrder(orderId: number, remarks?: string): Promise<{ success: boolean; message: string }> {
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${orderId}/expedite`, {
    method: 'POST',
    data: { remarks },
  });
}

/** 杂费项 (V2) */
export interface LandingCostFeeItem {
  name: string;
  amount: number;
}

/** 杂费分摊请求 (V2) */
export interface LandingCostAllocationRequest {
  fee_items: LandingCostFeeItem[];
  method: 'by_value' | 'by_quantity' | 'by_weight' | 'by_volume';
}

/** 分摊落地成本 (V2) */
export async function allocatePurchaseCosts(orderId: number, data: LandingCostAllocationRequest): Promise<any> {
  return apiRequest<any>(`/apps/kuaizhizao/purchase-orders/${orderId}/allocate-costs`, {
    method: 'POST',
    data,
  });
}

File: riveredge-backend/src/apps/kuaizhizao/schemas/purchase.py
"""
采购订单数据验证Schema

提供采购订单相关的Pydantic数据验证模型。

Author: Luigi Lu
Date: 2025-12-30
"""

from datetime import date, datetime
from typing import List, Optional
from decimal import Decimal

from pydantic import BaseModel, Field, ConfigDict, field_validator
from apps.kuaizhizao.services.document_action_policy.types import PurchaseOrderCapabilities

from apps.kuaizhizao.constants import DocumentStatus, ReviewStatus

# 导入BaseSchema用于兼容性
from core.schemas.base import BaseSchema


# === 采购订单 ===
class PurchaseOrderBase(BaseSchema):
    """采购订单基础Schema"""
    model_config = ConfigDict(from_attributes=True)

    order_code: Optional[str] = Field(None, max_length=50, description="订单编码")
    supplier_id: int = Field(..., description="供应商ID")
    supplier_name: str = Field(..., max_length=200, description="供应商名称")
    supplier_contact: Optional[str] = Field(None, max_length=100, description="供应商联系人")
    supplier_phone: Optional[str] = Field(None, max_length=20, description="供应商电话")
    buyer_id: Optional[int] = Field(None, description="归属采购员ID")
    buyer_name: Optional[str] = Field(None, max_length=100, description="归属采购员姓名")
    order_date: date = Field(..., description="订单日期")
    delivery_date: date = Field(..., description="要求到货日期")
    order_type: str = Field("标准采购", max_length=20, description="订单类型")
    total_quantity: Decimal = Field(default=Decimal(0), ge=0, description="总数量")
    total_amount: Decimal = Field(default=Decimal(0), ge=0, description="订单总金额")
    tax_rate: Decimal = Field(default=Decimal(0), ge=0, le=1, description="税率")
    tax_amount: Decimal = Field(default=Decimal(0), ge=0, description="税额")
    net_amount: Decimal = Field(default=Decimal(0), ge=0, description="净金额")
    currency: str = Field("CNY", max_length=10, description="币种")
    exchange_rate: Decimal = Field(default=Decimal(1), gt=0, description="汇率")
    status: str = Field(default=DocumentStatus.DRAFT.value, max_length=20, description="订单状态")
    source_type: Optional[str] = Field(None, max_length=50, description="来源类型")
    source_id: Optional[int] = Field(None, description="来源ID")
    notes: Optional[str] = Field(None, description="备注")
    attachments: Optional[List[dict]] = Field(None, description="附件列表")

    # 费用信息
    fee_details: Optional[List[dict]] = Field(None, description="费用明细 (JSON)")
    total_fee_amount: Decimal = Field(default=Decimal(0), ge=0, description="总费用金额")

    prepayment_amount: Optional[Decimal] = Field(None, ge=0, description="预付款金额")
    prepayment_bank_account_id: Optional[int] = Field(None, description="预付款银行账户ID")


class PurchaseOrderCreate(PurchaseOrderBase):
    """采购订单创建Schema"""
    order_code: Optional[str] = Field(None, max_length=50, description="订单编码")
    items: List["PurchaseOrderItemCreate"] = Field(..., description="订单明细")


class PurchaseOrderUpdate(PurchaseOrderBase):
    """采购订单更新Schema"""
    order_code: Optional[str] = Field(None, max_length=50, description="订单编码")
    items: Optional[List["PurchaseOrderItemUpdate"]] = Field(None, description="订单明细")
    attachments: Optional[List[dict]] = Field(None, description="附件列表")
    fee_details: Optional[List[dict]] = Field(None, description="费用明细 (JSON)")
    total_fee_amount: Optional[Decimal] = Field(None, ge=0, description="总费用金额")
    change_reason: Optional[str] = Field(None, description="变更原因（当已审核/确认后变更时必填）")


class PurchaseOrderResponse(PurchaseOrderBase):
    """采购订单响应Schema"""
    id: int = Field(..., description="订单ID")
    tenant_id: int = Field(..., description="租户ID")
    reviewer_id: Optional[int] = Field(None, description="审核人ID")
    reviewer_name: Optional[str] = Field(None, max_length=100, description="审核人姓名")
    review_time: Optional[datetime] = Field(None, description="审核时间")
    review_status: str = Field(default=ReviewStatus.PENDING.value, max_length=20, description="审核状态")
    review_remarks: Optional[str] = Field(None, description="审核备注")
    created_by: Optional[int] = Field(None, description="创建人ID")
    created_by_name: Optional[str] = Field(None, description="创建人姓名")
    updated_by: Optional[int] = Field(None, description="更新人ID")
    updated_by_name: Optional[str] = Field(None, description="更新人姓名")
    created_at: datetime = Field(..., description="创建时间")
    updated_at: datetime = Field(..., description="更新时间")
    items: List["PurchaseOrderItemResponse"] = Field(default_factory=list, description="订单明细")
    lifecycle: Optional[dict] = Field(None, description="生命周期（后端计算，供 UniLifecycleStepper 展示）")
    capabilities: Optional[PurchaseOrderCapabilities] = Field(
        None,
        description="业务态动作 capabilities（不含 RBAC，与 service 门禁一致）",
    )


class PurchaseOrderListResponse(PurchaseOrderResponse):
    """采购订单列表响应Schema"""
    items_count: Optional[int] = Field(None, description="订单明细条数（列表用）")
    downstream_push_progress: Optional[float] = Field(None, description="下推进度 0-100（列表用）")
    downstream_receipt_notice_codes: List[str] = Field(
        default_factory=list,
        description="下游收货通知单编码（列表 hover 用）",
    )
    downstream_purchase_receipt_codes: List[str] = Field(
        default_factory=list,
        description="下游采购入库单编码（列表 hover 用）",
    )
    received_total: Optional[Decimal] = Field(None, description="累计收货数量（列表用）")
    outstanding_total: Optional[Decimal] = Field(None, description="待收货数量（列表用）")
    receipt_progress: Optional[float] = Field(None, description="收货进度 0-100（列表用）")
    has_arrival_overdue: Optional[bool] = Field(
        None,
        description="是否存在逾期未关闭明细行（行级预警口径）",
    )


# === 采购订单明细 ===
class PurchaseOrderItemBase(BaseSchema):
    """采购订单明细基础Schema"""
    model_config = ConfigDict(from_attributes=True)

    material_id: int = Field(..., description="物料ID")
    material_code: str = Field(..., max_length=50, description="物料编码")
    material_name: str = Field(..., max_length=200, description="物料名称")
    material_spec: Optional[str] = Field(None, max_length=200, description="物料规格")
    ordered_quantity: Decimal = Field(..., gt=0, description="采购数量")
    unit: str = Field(..., max_length=20, description="单位")
    unit_price: Decimal = Field(..., ge=0, description="单价")
    total_price: Decimal = Field(..., ge=0, description="总价")
    price_settlement_status: Optional[str] = Field(None, max_length=20, description="定价状态")
    provisional_unit_price: Optional[Decimal] = Field(None, ge=0, description="暂估参考单价")
    received_quantity: Decimal = Field(default=Decimal(0), ge=0, description="已到货数量")
    outstanding_quantity: Decimal = Field(default=Decimal(0), ge=0, description="未到货数量")
    required_date: date = Field(..., description="要求到货日期")
    actual_delivery_date: Optional[date] = Field(None, description="实际到货日期")
    quality_requirements: Optional[str] = Field(None, description="质量要求")
    inspection_required: bool = Field(True, description="是否需要检验")
    source_type: Optional[str] = Field(None, max_length=50, description="来源类型")
    source_id: Optional[int] = Field(None, description="来源ID")
    demand_computation_item_id: Optional[int] = Field(None, description="需求计算明细ID")
    notes: Optional[str] = Field(None, description="备注")


class PurchaseOrderItemCreate(PurchaseOrderItemBase):
    """采购订单明细创建Schema"""
    pass


class PurchaseOrderItemUpdate(PurchaseOrderItemBase):
    """采购订单明细更新Schema"""
    material_id: Optional[int] = Field(None, description="物料ID")
    material_code: Optional[str] = Field(None, max_length=50, description="物料编码")
    material_name: Optional[str] = Field(None, max_length=200, description="物料名称")
    ordered_quantity: Optional[Decimal] = Field(None, gt=0, description="采购数量")


class PurchaseOrderItemResponse(PurchaseOrderItemBase):
    """采购订单明细响应Schema"""
    id: int = Field(..., description="明细ID")
    order_id: int = Field(..., description="订单ID")
    price_settled_at: Optional[datetime] = Field(None, description="定价时间")
    price_settled_by: Optional[int] = Field(None, description="定价人ID")
    # V2 落地成本增强
    landing_cost: Decimal = Field(default=Decimal(0), description="落地成本")
    additional_fees_details: Optional[List[dict]] = Field(None, description="费用明细")


# === 扩展响应 ===
class MaterialPriceHistoryItem(BaseModel):
    """物料成交价记录项"""
    order_id: int
    order_code: str
    order_date: date
    supplier_id: int
    supplier_name: str
    unit_price: Decimal
    currency: Optional[str] = None

class MaterialPriceHistoryResponse(BaseModel):
    """物料成交价查询响应"""
    material_id: int
    history_items: List[MaterialPriceHistoryItem] = Field(default_factory=list)
    average_price: Decimal = Field(default=Decimal(0))
    min_price: Decimal = Field(default=Decimal(0))
    max_price: Decimal = Field(default=Decimal(0))


class PurchaseTrackingNode(BaseModel):
    """单据追踪节点信息"""
    node_name: str
    status: str
    time: Optional[datetime] = None
    operator: Optional[str] = None
    detail: Optional[str] = None
    is_completed: bool = False
    is_warning: bool = False

class PurchaseTrackingResponse(BaseModel):
    """全链路追踪响应"""
    order_id: int
    order_code: str
    overall_progress: int = 0
    nodes: List[PurchaseTrackingNode] = Field(default_factory=list)


# === V2 增强：比价与分摊 ===
class PriceComparisonItem(BaseModel):
    """比价条目"""
    supplier_id: int
    supplier_name: str
    last_price: Decimal
    last_order_date: Optional[date]
    delivery_lead_time: int  # Days


class MaterialPriceComparison(BaseModel):
    """物料比价汇总"""
    material_id: int
    material_name: str
    material_code: Optional[str] = None
    comparison: List[PriceComparisonItem] = Field(default_factory=list)


class PriceComparisonResponse(BaseModel):
    """多物料比价响应"""
    results: List[MaterialPriceComparison]


class LandingCostFeeItem(BaseModel):
    name: str = Field(..., description="费用名称（如：运费、关税）")
    amount: Decimal = Field(..., description="金额")


class LandingCostAllocationRequest(BaseModel):
    fee_items: List[LandingCostFeeItem] = Field(..., description="待分摊杂费列表")
    method: str = Field("by_value", description="分摊方式：by_value, by_quantity, by_weight, by_volume")


class PurchaseOrderChangeResponse(BaseModel):
    id: int
    order_id: int
    change_type: str
    field_name: str
    old_value: Optional[str]
    new_value: Optional[str]
    reason: Optional[str]
    operator_id: Optional[int]
    operator_name: Optional[str]
    created_at: datetime

    class Config:
        from_attributes = True


# === 其他请求 ===
class PurchaseOrderApprove(BaseModel):
    approved: bool
    review_remarks: Optional[str] = None


class PurchaseOrderConfirm(BaseModel):
    confirm_remarks: Optional[str] = None


class PurchaseOrderListParams(BaseModel):
    supplier_id: Optional[int] = None
    status: Optional[str] = None
    review_status: Optional[str] = None
    order_date_from: Optional[date] = None
    order_date_to: Optional[date] = None
    delivery_date_from: Optional[date] = None
    delivery_date_to: Optional[date] = None
    created_start_date: Optional[date] = None
    created_end_date: Optional[date] = None
    order_code: Optional[str] = None
    keyword: Optional[str] = None
    order_by: Optional[str] = None
    pullable_only: Optional[bool] = None
    pull_target: Optional[str] = None
    include_items: bool = False
    skip: int = 0
    limit: int = 20


class PurchaseReceiptPullCandidate(BaseModel):
    """采购入库选单弹窗候选项（含明细数量汇总）"""
    id: int
    order_code: str
    supplier_name: Optional[str] = None
    status: str
    order_date: Optional[date] = None
    delivery_date: Optional[date] = None
    items_count: Optional[int] = None
    ordered_total: Decimal = Field(default=Decimal(0), description="采购数量合计")
    received_total: Decimal = Field(default=Decimal(0), description="已入库数量合计")
    outstanding_total: Decimal = Field(default=Decimal(0), description="未入库数量合计")
    pullable: bool = Field(..., description="是否可取单（与 capabilities.push_receipt 一致）")
    lifecycle: Optional[dict] = Field(None, description="生命周期（入库进度展示）")
    capabilities: Optional[PurchaseOrderCapabilities] = Field(
        None,
        description="业务态动作 capabilities（不含 RBAC，与 service 门禁一致）",
    )


class PurchaseReceiptPullCandidateListResponse(BaseModel):
    data: List[PurchaseReceiptPullCandidate]
    total: int
    success: bool = True
